### ISSUE TITLE
Handover Gray Box and Code Reviews to SKF

### ISSUE THREAD

**Issue by ThunderSon** [OWASP/wstg#592]
**What would you like to happen?**
Having a Code Review section, where a tester can communicate with the developers and have direct access to the source code is more descriptive than Gray Box, because actually having access to the source code is white box to begin with.

What do you think?

**Comment by kingthorin** [comment]
I'd rather leave code review for the code review guide (whether it's alive or not) or SKF.

**Comment by ThunderSon** [comment]
@RiieCco @blabla1337 Hi SKF team :heart: 
Would you be interested to host code review recommendations?

@kingthorin should we just remove the sections then that discuss code review?

**Comment by kingthorin** [comment]
I’m not against mentioning it, I just don’t think it’s the TG’s place to try to be authoritative in it.

**Comment by ThunderSon** [comment]
Something to keep in mind; sometimes checking the backend code is essential to validate the security (CRYP-04 is one example). Would that be considered stepping outside of the WSTG boundaries? Or would that be the actual test in that case? You can't actually see for example what encryption is being used and if it's securely implemented from outside.

**Comment by kingthorin** [comment]
Hmmm that's an interesting example, however, if you look through the Code Review section how/where do we draw the lines. There are some examples but they're java'centric. Maybe we don't care but the argument can easily be made by readers "You have java details why don't you have X details" where X is any language under the sun.

It also overlooks a ton of material like DB (at rest) encryption, drive encryption, S3 encryption, etc.

To me while it's good for the TG to acknowledge and even discuss some techniques outside L7 remote testing, if we try to make the guide all encompassing it'll never hit all the right points.

I'll open a separate issue for CRYP-04 it's got it's own issues (ex: rough English, a personal author/editor reference, etc.)

**Comment by RiieCco** [comment]
@ThunderSon, @kingthorin, we would love to host this!! 
How can we help? :-)

**Comment by ThunderSon** [comment]
That's easy :yum: 
From your side, you'd need to have the code review for the test cases. They can be more wide, or not specific to the WSTG.
From our side, we'll link which fits most to the tests.
Pretty sure I can help with that as well :)

Let me know of your timelines and see this through.

**Comment by ThunderSon** [comment]
@kingthorin after @RiieCco was going through the first 2 chapters, the gray box sections seemed to need reformulation, and not full removal.
Looking at this, I am worried that we could waste a lot of time on this, where as we can be restructuring a full test.

My suggestion comes now to close this issue and try to tackle things differently. This was too optimistic.

**Comment by kingthorin** [comment]
Alright that's fair. I'm okay with closing it.

@RiieCco if you have notes/thoughts on the content you did look at feel free to open other issues.

**Comment by patrickceg** [comment]
As a developer, my thoughts on code review or grey / white box:

* Anything where a human must _manually look at_ the configuration or source code to find vulnerabilities should go in a different guide (secure coding guides, hardening guides, etc) because staring at source code is not really a "test"
    * I don't think there is a case where you can have a vulnerability found by staring at the code, but there's no way to write a test to exploit it. There are of course exploits that cannot be run _reliably_ including exploitation of race conditions. However, if the probability of hitting a vulnerable piece of code is zero, it probably means that path is dead code (which now just goes in to software implementation guidelines).
* If there exists an _automated static analysis tool_ to find the potentially dangerous code pattern that could be a vulnerability (example: Static analysis including SpotBugs to find null dereference in back-ends, OWASP Dependency Check to check library version numbers), then it is a test and should go in this guide
* As a separate topic, *samples of vulnerable code or configurations* can still go in this guide on a case-by-case basis, because that would be required if a system fails a test and the tester + developers must collaborate to find what the cause is / is the failure a false positive.

I agree with the other comments that enforcing some criteria like this could result in maybe tens of pages of content needing to be replaced by references to secure coding / configuration guides, or even being removed completely from wstg if the only way to find a vulnerability happens to be black box test or stare at code manually with nothing in between.

(I've just started back up on side projects so this is my first post to this Github in easily months)

I wasn't part of the original discussion, but I wouldn't mind closing this task just because we discovered it's too broad.
I read we have multiple agreements that we would decrease the quality of the guide if we went in and blindly removed every "Grey Box Testing" with a code sample. We'd have to look at articles case-by-case, which means we need separate tasks.

As a note, if we decide to make tasks to define how we handle reviewing code, another article we need to look at is format strings <https://github.com/OWASP/wstg/blob/master/document/4-Web_Application_Security_Testing/07-Input_Validation_Testing/13-Testing_for_Format_String_Injection.md> where there's a manual code review section as well as some code samples.
